Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,332 @@
+* `o₁ = o₂` and `a ≤ b`
+* `o₁ < o₂` and `a ≤ veblenWith f o₂ b`
+* `o₁ > o₂` and `veblenWith f o₁ a ≤ b` -/
+theorem veblenWith_le_veblenWith_iff :
+    veblenWith f o₁ a ≤ veblenWith f o₂ b ↔
+      o₁ = o₂ ∧ a ≤ b ∨ o₁ < o₂ ∧ a ≤ veblenWith f o₂ b ∨ o₂ < o₁ ∧ veblenWith f o₁ a ≤ b := by
+  rw [← not_lt, ← cmp_eq_gt_iff, cmp_veblenWith_eq hf]
+  aesop (add simp [not_lt_of_le, lt_asymm])
+
+/-- `veblenWith f o₁ a = veblenWith f o₂ b` iff one of the following holds:
+* `o₁ = o₂` and `a = b`
+* `o₁ < o₂` and `a = veblenWith f o₂ b`
+* `o₁ > o₂` and `veblenWith f o₁ a = b` -/
+theorem veblenWith_eq_veblenWith_iff :
+    veblenWith f o₁ a = veblenWith f o₂ b ↔
+      o₁ = o₂ ∧ a = b ∨ o₁ < o₂ ∧ a = veblenWith f o₂ b ∨ o₂ < o₁ ∧ veblenWith f o₁ a = b := by
+  rw [← cmp_eq_eq_iff, cmp_veblenWith_eq hf]
+  aesop (add simp lt_asymm)
+
+end veblenWith
+
+/-! ### Veblen function -/
+
+section veblen
+
+/-- `veblen o` is the `o`-th function in the Veblen hierarchy starting with `ω ^ ·`. That is:
+
+- `veblen 0 a = ω ^ a`.
+- `veblen a` enumerates the fixed points of `veblen b` for `b < a` when `a ≠ 0`.
+-/
+@[pp_nodot]
+def veblen : Ordinal.{u} → Ordinal.{u} → Ordinal.{u} :=
+  veblenWith (ω ^ ·)
+
+@[simp]
+theorem veblen_zero : veblen 0 = fun a ↦ ω ^ a := by
+  rw [veblen, veblenWith_zero]
+
+theorem veblen_zero_apply (a : Ordinal) : veblen 0 a = ω ^ a := by
+  rw [veblen_zero]
+
+theorem veblen_of_ne_zero (h : o ≠ 0) : veblen o = derivFamily fun x : Set.Iio o ↦ veblen x.1 :=
+  veblenWith_of_ne_zero _ h
+
+theorem isNormal_veblen (o : Ordinal) : IsNormal (veblen o) :=
+  (isNormal_opow one_lt_omega0).veblenWith o
+
+theorem veblen_veblen_of_lt (h : o₁ < o₂) (a : Ordinal) : veblen o₁ (veblen o₂ a) = veblen o₂ a :=
+  veblenWith_veblenWith_of_lt (isNormal_opow one_lt_omega0) h a
+
+theorem veblen_succ (o : Ordinal) : veblen (Order.succ o) = deriv (veblen o) :=
+  veblenWith_succ (isNormal_opow one_lt_omega0) o
+
+theorem veblen_right_strictMono (o : Ordinal) : StrictMono (veblen o) :=
+  veblenWith_right_strictMono (isNormal_opow one_lt_omega0) o
+
+@[simp]
+theorem veblen_lt_veblen_iff_right : veblen o a < veblen o b ↔ a < b :=
+  veblenWith_lt_veblenWith_iff_right (isNormal_opow one_lt_omega0)
+
+@[simp]
+theorem veblen_le_veblen_iff_right : veblen o a ≤ veblen o b ↔ a ≤ b :=
+  veblenWith_le_veblenWith_iff_right (isNormal_opow one_lt_omega0)
+
+theorem veblen_injective (o : Ordinal) : Function.Injective (veblen o) :=
+  veblenWith_injective (isNormal_opow one_lt_omega0) o
+
+@[simp]
+theorem veblen_inj : veblen o a = veblen o b ↔ a = b :=
+  (veblen_injective o).eq_iff
+
+theorem right_le_veblen (a b : Ordinal) : b ≤ veblen a b :=
+  right_le_veblenWith (isNormal_opow one_lt_omega0) a b
+
+theorem veblen_left_monotone (o : Ordinal) : Monotone (veblen · o) :=
+  veblenWith_left_monotone (isNormal_opow one_lt_omega0) o
+
+@[simp]
+theorem veblen_pos (a b : Ordinal) : 0 < veblen a b :=
```

Comment:
```suggestion
theorem veblen_pos (o a : Ordinal) : 0 < veblen o a :=
```
Actually, I would even make the arguments implicit (and similarly for `veblenWith_pos`)